### PR TITLE
Feat  add support for sending outgoing messages

### PR DESCRIPTION
**Pull request by changangus**
## Description

App connects to websocket server on start. Support for sending outgoing messages to the websocket server on "enter" keypress event and form submit was added. A sendMessage function was created to bind both this function and the addMessageToChatHistory function to the events using a single callback. 

## Spec

See Issue: [FSA21V2-63](https://sparkbox.atlassian.net/browse/FSA21V2-63)

## Validation

- [ ] This PR has code changes, and our linters still pass.
- [ ] This PR has new code, so new tests were added or updated, and they pass.

### To Validate

1. Make sure all PR Checks have passed.
2. Pull down all related branches.
3. Run `npm run lint` - make sure there are no linting errors
4. Run `npm test`  - make sure all tests pass
5. Run `npm run dev:server`
6. Run `npm run dev`
7. Open app - there should be no alerts or error messages
8. Send message - there should be no alerts or error messages

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  "configurations": [
+    {
+      "type": "pwa-chrome",
+      "request": "launch",
+      "name": "Launch Chrome against localhost",
+      "url": "http://localhost:8080",
+      "webRoot": "${workspaceFolder}"
+    }
+  ]
+}
```
Comment (gtempus) [inline .vscode/launch.json]: You may not have intended to add this. On the other hand, maybe you did? 😄 
If you did want to add IDE configs then I would suggest creating a separate PR since it doesn't really relate to this PR.

File: server/package-lock.json [hunk 2]
```diff
@@ -3616,6 +3616,7 @@
       "version": "2.3.2",
       "resolved": "https://registry.npmjs.org/fsevents/-/fsevents-2.3.2.tgz",
       "integrity": "sha512-xiqMQR4xAeHTuB9uWm+fFRcIOgKBMiOBP+eXiyT7jsgVCq1bkVygt00oASowB7EdtpOHaaPgKt812P9ab+DDKA==",
+      "dev": true,
```
Comment (gtempus) [inline server/package-lock.json:3619]: This relates to an earlier question that I had about node/npm versions. I recommend that you folks agree on versioning and capture that in the `package.json`. Again, once your versions align, this might be better placed in a separate PR.

File: src/js/app.js [hunk 3]
```diff
@@ -1,8 +1,10 @@
...
+import { setCustomWebSocketInstance } from './chatService';
 import { onEnterPress, onFormSubmit } from './eventHandlers';
+import sendMessage from './sendMessage';
 
-export default function main() {
-  onEnterPress(addMessageToChatHistory);
-  onFormSubmit(addMessageToChatHistory);
+export default async function main() {
+  onEnterPress(sendMessage);
+  onFormSubmit(sendMessage);
```
Comment (gtempus) [inline src/js/app.js:8]: Love how this reads, @changangus! 💯

File: src/js/app.js [hunk 4]
```diff
@@ -1,8 +1,10 @@
...
+  await setCustomWebSocketInstance(new WebSocket('ws://localhost:3000'));
```
Comment (gtempus) [inline src/js/app.js:9]: A couple of things here that you folks may want to talk about. It's mostly around error handling and detection.

On a hunch, I did this terrible thing:
```suggestion
  await setCustomWebSocketInstance({});
```
Not surprisingly, the promise is never settled. The app continues cranking along for the user but chaos is happening under the hood:
- Websocket errors are reported to the console. 👍 
- The message textarea doesn't clear but the history is added to the UI. As a user I am annoyed that my message isn't cleared. More importantly, I _think_ my message is getting through since my message is posted to history. 🤔 

Now, I'm just getting curious. So I do this thing:
```suggestion
  await setCustomWebSocketInstance(new WebSocket('ws://localhosts:3000'));
```
Ack! Lots of good error output but still no feedback for the user. Interestingly, from the user perspective, my message in the textarea is cleared. Otherwise, my chat is never sent but it shows in chat history. So, it looks very convincing that the app is working. 😬 

This is fun stuff! Don't get discouraged. Talk it over with the team and make sure the behavior is spec'd out. These things don't all have to be _fixed_ but it should be shared and allow the PM to give guidance.

File: src/js/app.spec.js [hunk 5]
```diff
@@ -2,20 +2,20 @@
...
+    expect(onEnterPress).toHaveBeenCalledWith(sendMessage);
   });
 
-  it('Runs onFormSubmit with addMessageToChatHistory as a parameter', () => {
+  it('Runs onFormSubmit with sendMessage as a parameter', () => {
     main();
-    expect(onFormSubmit).toHaveBeenCalledWith(addMessageToChatHistory);
+    expect(onFormSubmit).toHaveBeenCalledWith(sendMessage);
   });
 });
```
Comment (gtempus) [inline src/js/app.spec.js:21]: This one is tough. On one hand, we exercise the collaborators of this module. On the other hand, it locks down what `app.js` does. So, i think, in this app's instance we should be very picky with our declarative api. I **love** a unified `sendMessage`. Maybe the same thing can be done with `onFormSubmit` and `onEnterPress` since it _feels_ kinda low level based on the chosen words (_enter_ and `formSubmit`)? 🤷 

What am I trying to say? Just keep this stuff in mind. I don't see any required changes, just things to consider. Be sensitive to pain points in your architecture.

What I do love is that this module's only job is to get collaborators together! Nice work!

File: src/js/chatService.spec.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    dispatchOutgoingMessage('Test');
+
+    expect(mockWebSocket.send).toHaveBeenCalled();
+  });
+  it('Resolves with instance', () => {
+    const mockWebSocket = {
+      onopen: () => {},
+      onerror: () => {},
+    };
```
Comment (gtempus) [inline src/js/chatService.spec.js:19]: As we start to see an API getting exposed it's good to realize that we don't own this. `send`, `onopen`, and `onerror` are WebSocket concepts. Again, just something to be mindful of. Even though I messed up the websockets initialization in my above comments/experiment, these tests continued to pass.

File: src/js/sendMessage.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import { dispatchOutgoingMessage } from './chatService';
+import { addMessageToChatHistory } from './chatWindow';
+
+export default function sendMessage(message) {
+  addMessageToChatHistory(message);
+  dispatchOutgoingMessage(message);
+}
```
Comment (gtempus) [inline src/js/sendMessage.js:7]: I love this abstraction! Well done, @changangus! 💪 

Some things that came up from my experimenting with websockets in this review:

- What if a message dispatch fails? This code already posts the message to history. The proverbial horse is already out of the barn. 🐴 
- What if the chat history fails? Should we still try to send the message? 

Work with your team to decide on this kind of behavior. At the very least, the issues will be surfaced and can be addressed now or later.

File: src/js/chatService.spec.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import { dispatchOutgoingMessage, setCustomWebSocketInstance } from './chatService';
+
+describe('chatService.js', () => {
```
Comment (gtempus) [inline src/js/chatService.spec.js:3]: i think we should add child `describe` blocks here. One for the creation/instantiation and the other for the behavior around `send`

**Review by gtempus — CHANGES_REQUESTED** [review]
Great PR, @changangus! I hope I provided some additional insight and gave you some interesting things to think about.

**Comment by gtempus** [comment]
Ooops, forgot. Please add some validation steps, too. Thanks, @changangus!

File: src/js/chatService.spec.js (hunk 8 above)
Comment (auaruss) [reply src/js/chatService.spec.js:3]: Note: I refactored this file in https://github.com/sparkbox/apprenticeship-chat-messenger/pull/10

Anyone can feel free to pull those changes into this PR (or do it a different way) if it makes more sense to y'all.

File: src/js/chatService.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+/* eslint-disable no-alert */
```
Comment (gtempus) [inline src/js/chatService.js]: I think this got left in my mistake. I don't see `alert` being used in this module.

File: src/js/index.js [hunk 10]
```diff
@@ -1,3 +1,7 @@
+/* eslint-disable no-alert */
 import main from './app';
 
-main();
+main().catch((error) => {
+  alert('Server Connection Failed!');
+  console.log(error);
+});
```
Comment (gtempus) [inline src/js/index.js:7]: Error handling does tend to make things messier. I think this is totally fine this early in the project. It can be addressed once the team decides how the behavior should be.

**Review by gtempus — CHANGES_REQUESTED** [review]
Looks good to me, @changangus. Looks like just one linter directive got left in my mistake.

**Review by gtempus — APPROVED** [review]
Thanks, @changangus! Approved!
